// memshared/lib/commands/pubsub.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _1 = require("../");
var subscriptions = {};
var masterSubscriptions = {};
/*
 * SUBSCRIBE
 * Subscribe to one channel, with the provided callback
 */
function subscribe(topic, callback) {
    if (!_1.isMasterNode()) {
        if (!subscriptions[topic]) {
            subscriptions[topic] = [];
        }
        subscriptions[topic].push(callback);
        _1.store.dispatch("subscribe", undefined, topic, _1.getProcessId());
    }
    else {
        if (!masterSubscriptions[topic]) {
            masterSubscriptions[topic] = [];
        }
        // "callback" is actually the process id here.
        masterSubscriptions[topic].push(callback);
    }
}
exports.subscribe = subscribe;
/*
 * UNSUBSCRIBE
 * Unsubscribe from one channel. Callback is optional.
 */
function unsubscribe(topic, callback) {
    if (!_1.isMasterNode()) {
        var hasCallback = (callback !== undefined);
        if (subscriptions[topic]) {
            if (hasCallback) {
                var index = subscriptions[topic].indexOf(callback);
                if (index !== -1) {
                    subscriptions[topic].splice(index, 1);
                }
            }
            else {
                delete subscriptions[topic];
            }
        }
        _1.store.dispatch("unsubscribe", undefined, topic, hasCallback, _1.getProcessId());
    }
    else {
        var hasCallback = callback;
        var processId = arguments[2];
        if (hasCallback) {
            var index = masterSubscriptions[topic].indexOf(processId);
            if (index !== -1) {
                masterSubscriptions[topic].splice(index, 1);
            }
        }
        else {
            delete masterSubscriptions[topic];
        }
    }
}
exports.unsubscribe = unsubscribe;
/*
 * PUBLISH channel message
 * Publish a message to an specific channel
 */
function publish(topic, message, isDispatching) {
    if (isDispatching === void 0) { isDispatching = true; }
    if (!_1.isMasterNode()) {
        if (isDispatching) {
            _1.store.dispatch("publish", undefined, topic, message);
        }
        else {
            subscriptions[topic].forEach(function (c) { return c(message); });
        }
    }
    else {
        if (masterSubscriptions[topic]) {
            masterSubscriptions[topic].forEach(function (processId) {
                var data = {
                    cmd: "publish",
                    args: [topic, message, false],
                    pubsub: true
                };
                if (_1.pm2) {
                    _1.pm2.sendDataToProcessId({
                        type: 'memshared',
                        data: data,
                        id: processId,
                        topic: 'memshared'
                    }, function (err, res) {
                        if (err)
                            return console.error("memshared: couldn't send message to worker.");
                    });
                }
                else {
                    _1.getProcessById(processId).send(data);
                }
            });
        }
    }
}
exports.publish = publish;
/*
 * PUBSUB channel
 * List the processes subscribing to this topic.
 */
function pubsub(topic, callback) {
    if (!_1.isMasterNode()) {
        _1.store.dispatch("pubsub", callback, topic);
    }
    else {
        return masterSubscriptions[topic];
    }
}
exports.pubsub = pubsub;
