// memshared/lib/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var cluster = require("cluster");
var commands = require("./commands");
var Store_1 = require("./Store");
exports.Store = Store_1.Store;
exports.store = new Store_1.Store();
var processesById = {};
function masterHandleIncomingMessage(processId, message) {
    if (!processMasterMessage(message)) {
        return;
    }
    // send result back to worker
    processesById[processId].send(message);
}
function workerHandleIncomingMessage(message) {
    if (message.topic === "memshared") {
        message = message.data;
    }
    if (!message || !commands[message.cmd]) {
        return;
    }
    if (message.messageId) {
        exports.store.consume(message);
    }
    else if (message.pubsub) {
        commands[message.cmd].apply(undefined, message.args);
    }
}
if (isMasterNode()) {
    // Setup existing workers
    Object.keys(cluster.workers).forEach(function (workerId) {
        registerProcess(cluster.workers[workerId].process);
    });
    // Listen for new workers to setup
    cluster.on("fork", function (worker) { return registerProcess(worker.process); });
    // Be notified when worker processes die.
    cluster.on('exit', function (worker, code, signal) {
        delete processesById[worker.process.pid];
    });
}
else {
    process.on("message", workerHandleIncomingMessage);
}
function isMasterNode() {
    return (!process.send);
}
exports.isMasterNode = isMasterNode;
function getProcessId() {
    return process.env.pm_id || process.pid;
}
exports.getProcessId = getProcessId;
function getProcessById(processId) {
    return processesById[processId];
}
exports.getProcessById = getProcessById;
function registerProcess(childProcess) {
    processesById[childProcess.pid] = childProcess;
    childProcess.on("message", function (message) { return masterHandleIncomingMessage(childProcess.pid, message); });
}
exports.registerProcess = registerProcess;
function processMasterMessage(message) {
    if (!message || !commands[message.cmd]) {
        return false;
    }
    // run command on master process
    try {
        message.result = commands[message.cmd].apply(undefined, message.args);
    }
    catch (e) {
        message.error = e.message;
    }
    // delete irrelevant data to send back to the worker
    delete message['args'];
    return true;
}
exports.processMasterMessage = processMasterMessage;
function setupPM2LaunchBus(mod) {
    exports.pm2 = mod;
    exports.pm2.launchBus(function (err, bus) {
        bus.on('memshared', function (packet) {
            var message = packet.data;
            var processId = packet.process.pm_id;
            if (!processMasterMessage(message)) {
                return;
            }
            exports.pm2.sendDataToProcessId({
                type: 'memshared',
                data: message,
                id: processId,
                topic: 'memshared'
            }, function (err, res) {
                if (err)
                    console.error("memshared: couldn't send message to worker.");
            });
        });
    });
}
exports.setupPM2LaunchBus = setupPM2LaunchBus;
function setup(data) {
    Object.assign(exports.store, data);
}
exports.setup = setup;
//
// Export commands
//
__export(require("./commands"));
