// memshared/lib/Store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _1 = require("./");
;
var messageId = 0;
var Store = (function () {
    function Store() {
        this.$callbacks = {};
    }
    Store.prototype.dispatch = function (cmd, callback) {
        var args = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            args[_i - 2] = arguments[_i];
        }
        var msg = this.buildMessage.apply(this, [cmd].concat(args));
        // callback when the worker receives back the final result
        if (callback) {
            this.$callbacks[msg.messageId] = callback;
        }
        if (process.env.pm_id === undefined) {
            // send command to be executed by the master node
            process.send(msg);
        }
        else {
            // PM2: send command to PM2's launchBus
            // (http://pm2.keymetrics.io/docs/usage/pm2-api/#send-message-to-process)
            process.send({
                type: "memshared",
                topic: "memshared",
                data: msg,
            });
        }
    };
    Store.prototype.consume = function (message) {
        if (this.$callbacks[message.messageId]) {
            // dispatch callback
            this.$callbacks[message.messageId](message.error, message.result);
            // cleanup
            delete this.$callbacks[message.messageId];
        }
    };
    Store.prototype.buildMessage = function (command) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        return {
            messageId: _1.getProcessId() + ":" + messageId++,
            cmd: command,
            args: args
        };
    };
    return Store;
}());
exports.Store = Store;
