// colyseus/lib/serializer/FossilDeltaSerializer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schema_1 = require("@colyseus/schema");
const fossilDelta = require("fossil-delta");
const msgpack = require("notepack.io");
const Protocol_1 = require("../Protocol");
const jsonPatch = require("fast-json-patch"); // this is only used for debugging patches
const Debug_1 = require("../Debug");
class FossilDeltaSerializer {
    constructor() {
        this.id = 'fossil-delta';
    }
    reset(newState) {
        this.previousState = newState;
        this.previousStateEncoded = msgpack.encode(this.previousState);
    }
    getFullState(client) {
        return this.previousStateEncoded;
    }
    applyPatches(clients, previousState) {
        const hasChanged = this.hasChanged(previousState);
        if (hasChanged) {
            let numClients = clients.length;
            while (numClients--) {
                const client = clients[numClients];
                Protocol_1.send[Protocol_1.Protocol.ROOM_STATE_PATCH](client, this.patches);
            }
        }
        return hasChanged;
    }
    hasChanged(newState) {
        const currentState = newState;
        let changed = false;
        let currentStateEncoded;
        /**
         * allow optimized state changes when using `Schema` class.
         */
        if (newState instanceof schema_1.Schema) {
            if (newState.$changed) {
                changed = true;
                currentStateEncoded = msgpack.encode(currentState);
            }
        }
        else {
            currentStateEncoded = msgpack.encode(currentState);
            changed = !currentStateEncoded.equals(this.previousStateEncoded);
        }
        if (changed) {
            this.patches = fossilDelta.create(this.previousStateEncoded, currentStateEncoded);
            //
            // debugging
            //
            if (Debug_1.debugPatch.enabled) {
                Debug_1.debugPatch('%d bytes, %j', this.patches.length, jsonPatch.compare(msgpack.decode(this.previousStateEncoded), currentState));
            }
            this.previousState = currentState;
            this.previousStateEncoded = currentStateEncoded;
        }
        return changed;
    }
}
exports.FossilDeltaSerializer = FossilDeltaSerializer;
